File: SortingAlgorithms/Insertion Sort/insertionsort.c
/**
 * Author: Ao Wang
 * Date: 08/18/19
 * Description: Insertion Sort Algorithm
 */

#include <stdio.h>

void insertionSort(int nums [], int len);
void swap(int nums [], int a, int b);
void print(int nums [], int len);

int main(){
    int nums [] = { 0, -1, 100, 5, -99, 85, 90, 3 };
    int len = sizeof(nums)/sizeof(nums[0]);
    printf("Unsorted: ");
    print(nums, len);
    printf("Sorted: ");
    insertionSort(nums, len);
    print(nums, len);
}

void insertionSort(int nums [], int len){
    for(int i = 1; i < len; i++){
        int j = i;
        while(j > 0 && nums[j - 1] > nums[j]){
            swap(nums, j - 1, j);
            j--;
        }
    }
}

void swap(int nums [], int a, int b){
    int temp = nums[a];
    nums[a] = nums[b];
    nums[b] = temp;
}

void print(int nums [], int len){
    printf("[");
    for(int i = 0; i < len - 1; i++){
        printf("%d, ", nums[i]);
    }
    printf("%d]", nums[len - 1]);
    printf("\n");
}

File: SortingAlgorithms/Insertion Sort/InsertionSort.py
"""
Author: Ao Wang
Date: 08/18/19
Description: Insertion Sort Algorithm
"""

def swap(nums, a, b):
    temp = nums[a]
    nums[a] = nums[b]
    nums[b] = temp

def insertion_sort(nums):
    for i in range(1, len(nums)):
        j = i
        while j > 0 and nums[j - 1] > nums[j]:
            swap(nums, j - 1, j)
            j -= 1
    return nums

nums = [0, -1, 100, 5, -99, 85, 90, 3]
print("Unsorted: " + str(nums))
insertion_sort(nums)
print("Sorted: " + str(insertion_sort(nums)))
